compiler/fir/semantics/src/org/jetbrains/kotlin/fir/resolve/dfa/LogicSystem.kt

```diff
@@ -126,7 +117,7 @@ abstract class LogicSystem<FLOW : Flow>(protected val context: ConeInferenceCont
         op: (Collection<Set<ConeKotlinType>>) -> MutableSet<ConeKotlinType>
     ): MutableTypeStatement {
         require(statements.isNotEmpty())
-        statements.singleOrNull()?.let { return it as MutableTypeStatement }
+        statements.singleOrNull()?.let { return it.asMutableStatement() }
         val variable = statements.first().variable
         assert(statements.all { it.variable == variable })
         val exactType = op.invoke(statements.map { it.exactType })
```

Comment:
Fun fact: the implementation of `or` and `and`, and the assumption that `and` is equivalent to `+=`, are wrong, because right now `!(x !is A && x !is B)` is equivalent to `x is A && x is B`. This probably doesn't matter for any real code though.